### PR TITLE
Allow Quantity to be initialized with a Distribution

### PR DESCRIPTION
**Pull request by mhvk**
This is really a spin-off from trying to get a general `Masked` class to work in #11127, where I wanted to ensuring that `Quantity(masked, unit)` gave a `MaskedQuantity`. But the changes I made suggested  `Quantity(distribution, unit)` should also give a `QuantityDistribution` - which indeed worked! I split the required changes out of #11127, since it is likely much easier to review here (and being able to do `distribution * unit` is nice!).

### CODE REVIEW

File: astropy/units/quantity.py
Comment (mhvk) [inline astropy/units/quantity.py:376]: Note: this could also be `subok=subok` so that `Quantity` initialization itself always returns a pure `Quantity` - on the other hand, it is not clear that getting a Quantity with a structured dtype out of a Distribution is useful for anything.

But I can change this if requested - most important I feel is that `distribution * unit` gives a `QuantityDistribution`.

**Comment by mhvk** [comment]
@eteq, @astrofrog - would one of you have time to review this (or suggest someone else)? I'd like to try to get the `Masked` PR in a state where it touches as little else as possible, so that it is easier to review... And the suggestion here seems good on its own.

**Comment by mhvk** [comment]
@eteq, @astrofrog - Gentle ping - quite simple PR and review would be nice...

**Review by astrofrog — COMMENTED** [review]
I'll defer to @eteq for a proper review, but reading over the docs I realized we never actually spell out what ``QuantityDistribution`` is, do we?

File: astropy/uncertainty/core.py
Comment (nstarman) [inline astropy/uncertainty/core.py:283]: Nitpick and feel free to ignore. I'm just thinking of McCabe complexity. I endeavor to structure code to minimize the complexity score and reducing the number of return statements is an easy couple points.

File: astropy/uncertainty/tests/test_distribution.py
Comment (nstarman) [inline astropy/uncertainty/tests/test_distribution.py:31]: Is the point here that we're instantiating from a view object? or is there something special about the transpose?

File: astropy/units/quantity.py
Comment (nstarman) [reply astropy/units/quantity.py:378]: Does this potentially open a problem if doing NDData * unit?

**Review by nstarman — COMMENTED** [review]
Nifty 👍

File: astropy/uncertainty/tests/test_distribution.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_distribution.py:31]: Here, I am really just collecting the function tests that @eteq put in originally in a class...  I think the original reason is that a view as a new `.dtype` gets taken and one better be sure that that is possible.  It would be better to actually check that the output has the right shape and so on, but maybe I can be excused because this is existing code :smiley_cat:

File: astropy/units/quantity.py
Comment (mhvk) [reply astropy/units/quantity.py:378]: Since `NDData` is not an `ndarray` subclass, it will likely break regardless... (though perhaps `nddata` overrides `__mul__` - it is one piece of astropy I have not really interacted with...).

Comment (nstarman) [reply astropy/units/quantity.py:378]: It doesn't appear to. Just had a quick peak.

File: astropy/uncertainty/core.py
Comment (mhvk) [reply astropy/uncertainty/core.py]: Numpy's strange choice to allow one to omit `dtype` and just pass in a class remains a bit annoying... I've made it slightly simpler (and found a bug on the way, so good!).

File: astropy/uncertainty/tests/test_distribution.py
Comment (nstarman) [reply astropy/uncertainty/tests/test_distribution.py:31]: GitHub needs a :shipit: (shipit) reaction. The rocket just doesn't suffice.

File: astropy/units/quantity.py
Comment (nstarman) [reply astropy/units/quantity.py:378]: That might be a good PR though... ``NDData * unit = NDData(..., unit=old_unit * unit)``

File: astropy/uncertainty/tests/test_distribution.py
Comment (eteq) [inline astropy/uncertainty/tests/test_distribution.py:51]: This is really smooth! As an editorial note I can't help but think we need to get the `<<` syntax more widely known...

File: astropy/uncertainty/tests/test_distribution.py
Comment (eteq) [inline astropy/uncertainty/tests/test_distribution.py:436]: This is confusing to me. Is it that this works for a regular Angle but *not* one that uses `Distribution`? Or it doesn't work for *any* angle? If it's the former I think it's probably a bug (which we can fix later to not hold up this PR, but should make an issue and add a comment here) whereas if it's the latter this should probably go in the regular `Angle` tests.

File: docs/changes/uncertainty/11210.bugfix.rst
Comment (eteq) [inline docs/changes/uncertainty/11210.bugfix.rst:2]: Should this instead be
```suggestion
initialize ``QuantityDistribution``.  Hence, ``distribution << unit``
```
?

**Review by eteq — APPROVED** [review]
LGTM broad-brush!  I have a few minor inline comments, but nothing major!

I also noticed @astrofrog's https://github.com/astropy/astropy/pull/11210#pullrequestreview-615842789 - @astrofrog, were you thinking @mhvk should do that in this PR, or that we should do that separately?

**Comment by astrofrog** [comment]
Follow up :)

File: astropy/uncertainty/tests/test_distribution.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_distribution.py:51]: Indeed!  Though I think the multiplication remains more intuitive and for most purposes the extra copy really doesn't matter.

Comment (nstarman) [reply astropy/uncertainty/tests/test_distribution.py:51]: does ``pd <<= u.ct`` work in these instances to avoid a copy?

File: astropy/uncertainty/tests/test_distribution.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_distribution.py:436]: Indeed, it works for `Angle` but cannot very easily work for any `Distribution` (since it has a structured `dtype`) - also just don't quite know what to do with things that change the number of elements (say, `.view('f4')`) - in practice, this basically never happens...

Anyway, I'll add a comment.

File: astropy/uncertainty/tests/test_distribution.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_distribution.py:51]: `pd << u.ct` already does not make a copy (it calls `Quantity(pd, u.ct, copy=False)`). `pd <<= u.ct` does not do anything different, really, except that `pd` now points to the new `QuantityDistribution` object.

File: docs/changes/uncertainty/11210.bugfix.rst
Comment (mhvk) [reply docs/changes/uncertainty/11210.bugfix.rst:2]: Both did not work before, in the sense that one did not get a `QuantityDistribution` out (but instead a `Quantity` with a structured dtype, which is less than useful:
```
In [7]: Distribution(np.ones((10,10)))
Out[7]: 
NdarrayDistribution([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]]) with n_samples=10

In [8]: Distribution(np.ones((10,10))) * u.m
Out[8]: 
<Quantity [([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],),
           ([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],)] m>
```

With this PR:
```
In [3]: Distribution(np.ones((10,10))) * u.m
Out[3]: 
<QuantityDistribution [[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
           [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]] m with n_samples=10>
```

File: astropy/uncertainty/tests/test_distribution.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_distribution.py]: Now added a comment and restructured the .view code as well.

File: docs/changes/uncertainty/11210.bugfix.rst
Comment (mhvk) [reply docs/changes/uncertainty/11210.bugfix.rst:2]: Now clarified.

**Pull request merged**
